torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,232 @@
+  static store_type store_;
+  static std::unordered_map<int, std::vector<std::string>> waiting;
+  query_type qt;
+  recv_bytes<query_type>(socket, &qt, 1);
+  if (qt == query_type::set) {
+    std::string key = recv_string(socket);
+    std::vector<char> data = recv_vector<char>(socket);
+    store_[key] = std::move(data);
+    // On "set" wake up all of the processes that wait
+    // for keys already in the store
+    for (auto& process : waiting) {
+      auto& p_socket = process.first;
+      auto& p_keys = process.second;
+      if (check(store_, p_keys)) {
+        wake_up(p_socket);
+        waiting.erase(p_socket);
+      }
+    }
+    return false;
+  } else if (qt == query_type::get) {
+    std::string key = recv_string(socket);
+    std::vector<char> data = store_.at(key);
+    send_vector(socket, data);
+    return false;
+  } else if (qt == query_type::wait) {
+    size_type nargs;
+    recv_bytes<size_type>(socket, &nargs, 1);
+    std::vector<std::string> keys(nargs);
+    for (int i = 0; i < nargs; i++) {
+      keys[i] = recv_string(socket);
+    }
+    if (check(store_, keys)) {
+      wake_up(socket);
+    } else {
+      waiting[socket] = std::move(keys);
+    }
+    return false;
+  } else if (qt == query_type::finish) {
+    return true;
+  } else {
+    throw std::runtime_error("expected a query type");
+  }
+}
+
+static bool check(const store_type& store,
+                  const std::vector<std::string>& keys) {
+  for (const std::string& key : keys) {
+    if (store.find(key) == store.end())
+      return false;
+  }
+  return true;
+}
+
+static void wake_up(int socket) {
+  query_type res = query_type::stop_waiting;
+  send_bytes<query_type>(socket, &res, 2);
+}
+
+// TODO move these to ChannelUtils?
+/* send a string's length and data */
+void send_string(int socket, const std::string& str,
+                 bool more_data = false) {
+  size_type size = str.size();
+  send_bytes<size_type>(_socket, &size, 1, true);
+  send_bytes<char>(_socket, str.data(), size, more_data);
+}
+
+/* send a vector's length and data */
+template<typename T>
+void send_vector(int socket, const std::vector<T>& vec,
+                 bool more_data = false) {
+  size_type size = vec.size();
+  send_bytes<size_type>(_socket, &size, 1, true);
+  send_bytes<T>(_socket, vec.data(), size, more_data);
+}
+
+/* receive a string as sent in send_vector */
+std::string recv_string(int socket) {
+  size_type value_size;
+  recv_bytes<size_type>(socket, &value_size, 1);
+  std::vector<char> value(value_size);
+  recv_bytes<char>(socket, value.data(), value.size());
+  return std::string(value.data(), value.size());
+}
+
+/* receive a vector as sent in send_vector */
+std::vector<T> recv_vector(int socket) {
+  size_type value_size;
+  recv_bytes<size_type>(socket, &value_size, 1);
+  std::vector<char> value(value_size);
+  recv_bytes<char>(socket, value.data(), value.size());
+  return value;
+}
+
+void send_query(int socket, query_type query) {
```

Comment:
Could you just name it `send_value` and make it template so any type can be passed?